packages-exp/messaging-exp/src/index.sw.ts

```diff
@@ -15,16 +15,29 @@
  * limitations under the License.
  */
 
+import '@firebase/installations-exp';
+
+import { ERROR_FACTORY, ErrorCode } from './util/errors';
+
 import { FirebaseMessaging } from './interfaces/public-types';
+import { isSwSupported } from './api/isSupported';
 import { registerMessaging } from './helpers/register';
-import '@firebase/installations-exp';
 
 export { onBackgroundMessage, getMessaging } from './api';
+export { isSwSupported as isSupported } from './api/isSupported';
 
 declare module '@firebase/component' {
   interface NameServiceMapping {
     'messaging-exp': FirebaseMessaging;
   }
 }
 
+// Top-level 'await' requires 'module' option set to 'esnext' or 'system', and 'target' option
+// set to 'es2017' or higher. For compatibility, use async expression here.
+void (async () => {
+  if (!(await isSwSupported())) {
+    throw ERROR_FACTORY.create(ErrorCode.UNSUPPORTED_BROWSER);
```

Comment:
I think we should check `isSupported()` in the factory function like you do originally, so that developers have a chance to call `isSupported()` themselves and to not initialize messaging in unsupported environments without getting an exception.

That means we need 2 factories, one for the client and one for the sw.

Reply:
make sense. The solution here is problematic because devs can't handle before the error is thrown.